codegenerator/cli/templates/static/codegen/src/eventFetching/hypersync/HyperSync.res

```diff
@@ -316,28 +320,67 @@ module BlockData = {
         let logger = Logging.createChild(~params={"url": serverUrl})
         let delayMilliseconds = 100
         logger->Logging.childInfo(
-          `Block #${blockNumber->Int.toString} not found in HyperSync. HyperSync has multiple instances and it's possible that they drift independently slightly from the head. Indexing should continue correctly after retrying the query in ${delayMilliseconds->Int.toString}ms.`,
+          `Block #${fromBlock->Int.toString} not found in HyperSync. HyperSync has multiple instances and it's possible that they drift independently slightly from the head. Indexing should continue correctly after retrying the query in ${delayMilliseconds->Int.toString}ms.`,
         )
         await Time.resolvePromiseAfterDelay(~delayMilliseconds)
-        await queryBlockData(~serverUrl, ~blockNumber, ~logger)
+        await queryBlockData(~serverUrl, ~fromBlock, ~toBlock, ~logger)
+      }
+    | Some(res) =>
+      switch res->convertResponse {
+      | Error(_) as err => err
+      | Ok(datas) if res.nextBlock <= toBlock => {
+          let restRes = await queryBlockData(
+            ~serverUrl,
+            ~fromBlock=res.nextBlock,
+            ~toBlock,
+            ~logger,
+          )
+          restRes->Result.map(rest => datas->Array.concat(rest))
+        }
+      | Ok(_) as ok => ok
       }
-    | Some(res) => res->convertResponse->Result.map(res => res->Array.get(0))
     }
   }
 
   let queryBlockDataMulti = async (~serverUrl, ~blockNumbers, ~logger) => {
-    let res =
-      await blockNumbers
-      ->Array.map(blockNumber => queryBlockData(~blockNumber, ~serverUrl, ~logger))
-      ->Promise.all
-    res
-    ->Utils.Array.transposeResults
-    ->Result.map(Array.keepMap(_, v => v))
+    switch blockNumbers->Array.get(0) {
+    | None => Ok([])
+    | Some(firstBlock) => {
+        let fromBlock = ref(firstBlock)
+        let toBlock = ref(firstBlock)
+        let set = Utils.Set.make()
+        for idx in 0 to blockNumbers->Array.length - 1 {
+          let blockNumber = blockNumbers->Array.getUnsafe(idx)
+          if blockNumber < fromBlock.contents {
+            fromBlock := blockNumber
+          }
+          if blockNumber > toBlock.contents {
+            toBlock := blockNumber
+          }
+          set->Utils.Set.add(blockNumber)->ignore
+        }
+        let res = await queryBlockData(
+          ~fromBlock=fromBlock.contents,
+          ~toBlock=toBlock.contents,
+          ~serverUrl,
+          ~logger,
+        )
+        res->Result.map(datas => {
+          datas->Array.keep(data => set->Utils.Set.has(data.blockNumber))
+        })
+      }
+    }
```

Comment:
And it would probably be good to validate that all block numbers that are queried, are in fact returned